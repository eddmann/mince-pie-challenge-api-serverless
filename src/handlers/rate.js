/* @flow */

import type { Pie } from '../types';

import { notFound, badRequest, noContent, json } from '../helpers/http';
import { createHandler, strictHttpAuthentication } from '../helpers/handlers';

const avgRating = (pie: Pie): number => {
  const ratings = Object.values(pie.Ratings);
  return ratings.reduce((acc: number, rating: any) => acc + rating, 0) / ratings.length;
};

const isValid = (rating: number) => rating > 0 && rating < 6;

const rate = async ({ event, callback, userId, services: { db } }) => {
  const { id } = event.pathParameters || {};

  const pie = await db.get(id);

  if (!pie) {
    callback(undefined, notFound('Unable to find the specified mince pie.'));
    return;
  }

  const { rating } = json(event.body);

  if (!isValid(rating)) {
    callback(
      undefined,
      badRequest('Invalid request body', [{ name: 'rating', reason: 'You must supply a valid pie rating.' }]),
    );
    return;
  }

  pie.Ratings[userId] = rating;
  pie.AvgRating = avgRating(pie);

  const attrs = {
    UpdateExpression: 'SET Ratings = :Ratings, AvgRating = :AvgRating',
    ExpressionAttributeValues: { ':Ratings': pie.Ratings, ':AvgRating': pie.AvgRating },
    ReturnValues: 'ALL_NEW',
  };

  await db.update(pie.Id, attrs);

  callback(undefined, noContent());
};

export default createHandler(strictHttpAuthentication(rate));
