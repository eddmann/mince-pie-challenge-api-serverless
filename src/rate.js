/* @flow */

import * as db from './services/dynamoPieDB';
import { getUserIdFromToken } from './services/cognitoUserToken';
import rate from './handlers/rate';

export const handler = rate({ getUserIdFromToken, db });
