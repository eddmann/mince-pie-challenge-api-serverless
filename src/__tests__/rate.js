import rate from '../handlers/rate';

const response = () =>
  function cb(error, response) {
    if (error) fail(`Unexpected error: ${error}`);
    cb.response = response;
  };

it('requires an authenticated user', async () => {
  const services = { getUserIdFromToken: () => Promise.resolve() };
  const callback = response();

  await rate(services)({ headers: { Authorization: '' } }, {}, callback, undefined);

  expect(callback.response).toMatchSnapshot();
});

it('ensures only valid ratings are stored', async () => {
  const pie = {
    Id: '1',
    UserId: 'USER_ID',
    Name: 'Another Mince Pie',
    AvgRating: 0,
    PhotoUrl: 'http://photo.url',
    ThumbnailUrl: 'http://thumbnail.url',
    Ratings: {},
    AddedAt: 1528217691,
  };

  const services = {
    getUserIdFromToken: () => Promise.resolve('USER_ID'),
    db: {
      get: () => Promise.resolve(pie),
      update: jest.fn(),
    },
  };
  const body = JSON.stringify({ rating: 7 });
  const callback = response();

  await rate(services)({ headers: { Authorization: 'TOKEN' }, body }, {}, callback, undefined);

  expect(callback.response).toMatchSnapshot();
});

it('rates a pie', async () => {
  const pie = {
    Id: '1',
    UserId: 'USER_ID',
    Name: 'Another Mince Pie',
    AvgRating: 0,
    PhotoUrl: 'http://photo.url',
    ThumbnailUrl: 'http://thumbnail.url',
    Ratings: {},
    AddedAt: 1528217691,
  };

  const services = {
    getUserIdFromToken: () => Promise.resolve('USER_ID'),
    db: {
      get: () => Promise.resolve(pie),
      update: jest.fn(),
    },
  };
  const body = JSON.stringify({ rating: 5 });
  const callback = response();

  await rate(services)({ headers: { Authorization: 'TOKEN' }, body }, {}, callback, undefined);

  expect(callback.response.statusCode).toEqual(204);
  expect(services.db.update).toBeCalled();
  expect(services.db.update.mock.calls[0][0]).toEqual('1');
});
