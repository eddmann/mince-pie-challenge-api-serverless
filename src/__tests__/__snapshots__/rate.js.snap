// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ensures only valid ratings are stored 1`] = `
Object {
  "body": "{\\"title\\":\\"Bad Request\\",\\"detail\\":\\"Invalid request body\\",\\"errors\\":[{\\"name\\":\\"rating\\",\\"reason\\":\\"You must supply a valid pie rating.\\"}]}",
  "headers": Object {
    "Access-Control-Allow-Credentials": "*",
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/problem+json",
  },
  "statusCode": 400,
}
`;

exports[`requires an authenticated user 1`] = `
Object {
  "body": "{\\"title\\":\\"Forbidden\\",\\"detail\\":\\"Service requires authenticated user\\"}",
  "headers": Object {
    "Access-Control-Allow-Credentials": "*",
    "Access-Control-Allow-Origin": "*",
    "Content-Type": "application/problem+json",
  },
  "statusCode": 403,
}
`;
